[jit] improve error message on recursive class defs

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21842 [jit] improve error message on recursive class defs**
* #21841 [jit] improve recursive scripting error message

We don't support this yet, so it should error clearly.

Differential Revision: [D15857568](https://our.internmc.facebook.com/intern/diff/D15857568)

suo committed:
[jit] improve error message on recursive class defs

We don't support this yet, so it should error clearly.

suo committed:
Update on "[jit] improve error message on recursive class defs"

[jit] improve error message on recursive class defs

We don't support this yet, so it should error clearly.

gh-metadata: pytorch pytorch 21842 gh/suo/63/head

zdevito commented on test/test_jit.py:
```diff
@@ -16268,15 +16268,30 @@ def copy(self):
+        """
+        Recursive class types not yet supported. We should give a good error message.
+        """
+        with self.assertRaisesRegex(RuntimeError, "Recursive class attributes not yet supported"):
```
"Recursive class" is a bit vague as a concept can it be made more clear what is being objected to here?

zdevito reviewed (approved):
Yeah, good idea.

eellison commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -148,6 +148,21 @@ std::vector<std::shared_ptr<SugaredValue>> SimpleValue::asTuple(
+  // Recursively check contained types. We need to do this because a user may do
+  // A -> B -> A.
+  for (const auto& type : attrType->containedTypes()) {
+    if (isRecursive(classType, type)) {
```
The naive search here seems like it could get really inefficient with any complicated hierarchies

suo replied:
BFS would be faster given how types are structured. But like, how complicated are we talking lol—this search is linear in the size of the type DAG, so unless someone is making 100000-nested compound types I think we're ok.

suo replied:
I guess we're doing a second traversal if it's a tuple type, so if we have deep nested tuples we may need to add manually unrolling.

eellison replied:
I think this is exponential in the worst case...

eellison replied:
If you have `N` numbered nodes where each node is connected to all nodes with a lower number

suo committed:
Update on "[jit] improve error message on recursive class defs"

[jit] improve error message on recursive class defs

We don't support this yet, so it should error clearly.

gh-metadata: pytorch pytorch 21842 gh/suo/63/head

Krovatkin reviewed (approved):
:shipit:

suo committed:
Update on "[jit] improve error message on recursive class defs"

[jit] improve error message on recursive class defs

We don't support this yet, so it should error clearly.

gh-metadata: pytorch pytorch 21842 gh/suo/63/head

suo committed:
Update on "[jit] improve error message on recursive class defs"

[jit] improve error message on recursive class defs

We don't support this yet, so it should error clearly.

gh-metadata: pytorch pytorch 21842 gh/suo/63/head
